From 5603704d04f4ed43d62b36883073bd1157c9dd25 Mon Sep 17 00:00:00 2001
From: Howard Chu <hyc@openldap.org>
Date: Wed, 8 Jan 2020 22:22:14 +0000
Subject: [PATCH] ITS#9146 syncprov: fix sessionlog init

---
 servers/slapd/overlays/syncprov.c | 7 +------
 1 file changed, 1 insertion(+), 6 deletions(-)

diff --git a/servers/slapd/overlays/syncprov.c b/servers/slapd/overlays/syncprov.c
index c8e457c42..0e1f362f0 100644
--- a/servers/slapd/overlays/syncprov.c
+++ b/servers/slapd/overlays/syncprov.c
@@ -3079,12 +3079,7 @@ sp_cf_gen(ConfigArgs *c)
 		}
 		sl = si->si_logs;
 		if ( !sl ) {
-			sl = ch_malloc( sizeof( sessionlog ));
-			sl->sl_mincsn = NULL;
-			sl->sl_sids = NULL;
-			sl->sl_num = 0;
-			sl->sl_numcsns = 0;
-			sl->sl_head = sl->sl_tail = NULL;
+			sl = ch_calloc( 1, sizeof( sessionlog ));
 			ldap_pvt_thread_mutex_init( &sl->sl_mutex );
 			si->si_logs = sl;
 		}
-- 
2.20.1

