From 4e8f9b90e01d093ffe9818ebc9d44b0956cfd9d6 Mon Sep 17 00:00:00 2001
From: Howard Chu <hyc@openldap.org>
Date: Sat, 11 Jan 2020 04:24:54 +0000
Subject: [PATCH] ITS#9150 fix nosync FALSE config

---
 servers/slapd/back-mdb/config.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/servers/slapd/back-mdb/config.c b/servers/slapd/back-mdb/config.c
index 46cb59d1f..f1b86c07a 100644
--- a/servers/slapd/back-mdb/config.c
+++ b/servers/slapd/back-mdb/config.c
@@ -594,7 +594,7 @@ mdb_cf_gen( ConfigArgs *c )
 		if ( c->value_int )
 			mdb->mi_dbenv_flags |= MDB_NOSYNC;
 		else
-			mdb->mi_dbenv_flags ^= MDB_NOSYNC;
+			mdb->mi_dbenv_flags &= ~MDB_NOSYNC;
 		if ( mdb->mi_flags & MDB_IS_OPEN ) {
 			mdb_env_set_flags( mdb->mi_dbenv, MDB_NOSYNC,
 				c->value_int );
-- 
2.20.1

